File: src/frontend/static/js/character.js
document.addEventListener("DOMContentLoaded", async function () {
    try {
        const response = await fetch("http://127.0.0.1:8000/character");
        if (!response.ok) throw new Error("Failed to fetch character data");

        const data = await response.json();

        // Populate character information
        document.getElementById("char-name").textContent = data.name;
        document.getElementById("char-race").textContent = data.race;
        document.getElementById("char-class").textContent = data.class;
        document.getElementById("char-background").textContent = data.background;
        document.getElementById("char-hp").textContent = data.current_hit_points;
        document.getElementById("char-ac").textContent = data.armor_class;
        document.getElementById("char-gold").textContent = data.gold.toFixed(2);

        // Populate Traits
        const traitsList = document.getElementById("char-traits");
        traitsList.innerHTML = ""; // Clear previous entries

        if (data.traits && data.traits.length > 0) {
            data.traits.forEach(trait => {
                const listItem = document.createElement("li");
                listItem.textContent = trait;
                traitsList.appendChild(listItem);
            });
        } else {
            traitsList.innerHTML = "<li>No racial traits</li>";
        }

        // Populate Attributes
        const statsList = document.getElementById("char-stats");
        statsList.innerHTML = "";

        const attributes = [
            { name: "Strength", value: data.strength },
            { name: "Dexterity", value: data.dexterity },
            { name: "Constitution", value: data.constitution },
            { name: "Intelligence", value: data.intelligence },
            { name: "Wisdom", value: data.wisdom },
            { name: "Charisma", value: data.charisma }
        ];

        attributes.forEach(attr => {
            const listItem = document.createElement("li");
            listItem.textContent = `${attr.name}: ${attr.value}`;
            statsList.appendChild(listItem);
        });

        // Populate Saving Throws
        const savingThrowsList = document.getElementById("char-saving-throws");
        savingThrowsList.innerHTML = "";

        if (data.saving_throws && data.saving_throws.length > 0) {
            data.saving_throws.forEach(throwStat => {
                const listItem = document.createElement("li");
                listItem.textContent = throwStat;
                savingThrowsList.appendChild(listItem);
            });
        } else {
            savingThrowsList.innerHTML = "<li>No saving throw proficiencies</li>";
        }

        // Populate Proficiencies
        const proficienciesList = document.getElementById("char-proficiencies");
        proficienciesList.innerHTML = "";

        if (data.proficiencies && data.proficiencies.length > 0) {
            data.proficiencies.forEach(prof => {
                const listItem = document.createElement("li");
                listItem.textContent = prof;
                proficienciesList.appendChild(listItem);
            });
        } else {
            proficienciesList.innerHTML = "<li>No proficiencies</li>";
        }

        // Populate Equipment (Inventory)
        const inventoryList = document.getElementById("char-inventory");
        inventoryList.innerHTML = "";

        if (data.inventory && data.inventory.length > 0) {
            data.inventory.forEach(item => {
                const listItem = document.createElement("li");
                listItem.textContent = item;
                inventoryList.appendChild(listItem);
            });
        } else {
            inventoryList.innerHTML = "<li>No equipment</li>";
        }

    } catch (error) {
        console.error("Error loading character sheet:", error);
        document.getElementById("character-info").innerHTML = "<p>Error loading character data.</p>";
    }
});

File: src/backend/orchestrator/routes/character.py
"""Character creation endpoint configuration"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from src.backend.database.config import SessionLocal
from src.backend.database.models import Background, Character, CharacterClass, Race

router = APIRouter()

# Default character settings
DEFAULT_RACE = "Human"
DEFAULT_CLASS = "Ranger"
DEFAULT_BACKGROUND = "Folk Hero"


def get_db():
    """Dependency to get a new database session."""
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()


@router.get("/character")
def get_character(db: Session = Depends(get_db)):
    """Returns the character sheet details."""
    character = db.query(Character).first()
    if not character:
        raise HTTPException(status_code=404, detail="No character found")

    # Fetch related class, race, and background details
    race = db.query(Race).filter(Race.id == character.race_id).first()
    char_class = db.query(CharacterClass).filter(CharacterClass.id == character.class_id).first()
    background = db.query(Background).filter(Background.id == character.background_id).first()

    # Fetch racial traits
    traits = [trait.name for trait in race.traits] if race and race.traits else []

    # Fetch saving throws
    saving_throws = char_class.saving_throws if char_class and char_class.saving_throws else []

    # Fetch proficiencies from class and background
    class_proficiencies = (
        [prof.name for prof in char_class.proficiencies]
        if char_class and char_class.proficiencies
        else []
    )
    background_proficiencies = (
        [prof.name for prof in background.starting_proficiencies]
        if background and background.starting_proficiencies
        else []
    )

    # Combine all proficiencies, removing duplicates
    all_proficiencies = sorted(set(class_proficiencies + background_proficiencies))

    # Fetch inventory items
    inventory_list = [item.name for item in character.inventory] if character.inventory else []

    return {
        "name": character.name,
        "race": race.name if race else "Unknown",
        "class": char_class.name if char_class else "Unknown",
        "background": background.name if background else "Unknown",
        "current_hit_points": character.current_hit_points,
        "armor_class": character.armor_class,
        "gold": character.gold,
        "traits": traits,
        "strength": character.strength,
        "dexterity": character.dexterity,
        "constitution": character.constitution,
        "intelligence": character.intelligence,
        "wisdom": character.wisdom,
        "charisma": character.charisma,
        "saving_throws": saving_throws,
        "proficiencies": all_proficiencies,
        "inventory": inventory_list,
    }
